### ISSUE TITLE
Training mode

### ISSUE THREAD

**Issue by Fransje** [inexor-game/code#102]
Add a game mode in which the player is able to change almost every gameplay variable there is:
    - reload times
    - damage
    - what ammo you spawn with
    - pushback
    - how many added ammo at item pickup
    - blast radius for grenade and rocket explosions
    - speed of projectiles

This could be very useful when two players want to train their chaingun skills: just change the vars, so that every player spawns with 99999999 chaingun ammo and nothing else.
This might also be a nice way for professionals to test potential weapon balances (like lowering the rocket damage a bit).

**Comment by koraa** [comment]
+1

**Comment by Croydon** [comment]
That would be some kind of sandbox mode which sounds nice, on the other hand, writting cheats would become even easier as it is already. Could theoretically get implemented in coop mode.

**Comment by aschaeffer** [comment]
+1

**Comment by Fransje** [comment]
@Croydon I do not understand enough of the Sauer code to say something useful about the cheating risk, but it has always been open source, so whether there is a sandbox mode or not, cheating will always be possible. Maybe Inexor could try to implement a simple serverside cheat-detection? A few reload time, speed, jumpheight and edittoggle tests for example would hold back a good amount of non-malicious cheaters.

**Comment by IAmNotHanni** [comment]
@Fransje I agree that we'll have to improve the server-side anti cheat systems. Luckily we are in a very fast octree system so some cheat detection *could* become very easy for us. Checking health, ammo or the players speed is quite easy :)

**Comment by aschaeffer** [comment]
I like the idea.

* A nice HTML GUI with sliders would make it easy to adjust the training parameters
* We need to synchronize these parameters with the other clients
* The mode should be separate from coop-edit because it has nothing to do with editing
* It should be possible to set who can change the parameters (mastermode: all, only master, none)

The anti-cheat part shouldn't be discussed here, because you could write a cheat with and without this feature. In contrast it could lead to less cheats because currently it's the forbidden fruit which magically attracts cheaters. With an official coop-training mode cheating may become stale.

**Comment by Croydon** [comment]
@aschaeffer Agreed. Till now I also liked the idea that servers can register themselved on the masterserver with three levels:
0 = vanilla server
1 = modded server, without gameplay modifies
2 = modded gameplay

Level 2 servers should be marked that way somehow on the serverlist. Maybe also they shouldn't be listed via default, but should be easily blend in via the option menu.

When we would create now such a "training" mode, this mode would be likely getting used for all kind of modded gameplay servers. I like more superTux's (?) idea of (automatic) downloadable game modes (which comes very close to such a training mode), together with the server attribute of the modded level.

**Comment by Fohlen** [comment]
Refers to #339
